devito/finite_differences/derivative.py

```diff
@@ -405,7 +405,8 @@ def _eval_at(self, func):
             return self
         # For basic equation of the form f = Derivative(g, ...) we can just
         # compare staggering
-        if self.expr.staggered == func.staggered:
+        if self.expr.staggered == func.staggered and \
+           not (self.expr.is_Add or self.expr.is_Mul):
```

Comment:
what if it were a Pow or a trascendental function or something else ? this check isn't very neat